### ISSUE TITLE
testing and roll-out (to production) of code patch for Zombie issue 1744

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#1754]
This didn't get much traction in Slack, so posting over here as well:

On Nov 30's Architecture call, the proposed change to ARAX came up. The question came up if we had tested the demo queries before and after the change. I opined that there should not be any change, but admitted that we hadn't done the experiment. I asked if anyone had a tool to easily perform the experiment. Apparently there was not, but Chris created it here: https://github.com/NCATSTranslator/minihackathons/pull/368

Is anyone willing to perform the experiment so we can report?

**Comment by saramsey** [comment]
Per my Slack post, I am on it.

**Comment by saramsey** [comment]
My rough plan:

- [ ] start up the `arax-backup.rtx.ai` instance
- [ ] within the instance, start the `rtx1` container
- [ ] within the container, start `apache2`
- [ ] switch the `kg2` dev-area and `production` dev-area to the `production` branch
- [ ] start `RTX_OpenAPI_kg2`
- [ ] start `RTX_OpenAPI_production`
- [ ] run the workflow test and save the JSON results to a file `results-before.json`
- [ ] stop `RTX_OpenAPI_kg2`
- [ ] stop `RTX_OpenAPI_production`
- [ ] deploy PR 1748 patch into the `kg2` dev-area on top of the "production" branch
- [ ] deploy PR 1748 patch into the `production` dev-area on top of the "production" branch
- [ ] start `RTX_OpenAPI_kg2`
- [ ] start `RTX_OpenAPI_production`
- [ ] run the workflow test and save the JSON results to a file `results-after.json`
- [ ] diff the two files and report back in this issue if there are any discrepancies

**Comment by saramsey** [comment]
Pasting here the script (and associated query JSON file) that I used to test whether the local ARAX is running and if it is using the local RTX-KG2 API. I ran `run-curl.sh` once with the `RTX_OpenAPI_kg2` service running, and once when it was not running, and compared the JSON outputs from the two tests.

First, `run-curl.sh`
```
#!/usr/bin/env bash
curl -X POST "http://localhost:5000/api/arax/v1.2/query" \
     -H "accept: application/json" \
     -H "Content-Type: application/json" \
     -d @kg2-test.json 
```
Now, `kg2-test.json`:
```
{
    "message": {},
    "operations": 	{
	"actions": [
	    "add_qnode(ids=[CHEMBL.COMPOUND:CHEMBL112], key=n00)",
	    "add_qnode(ids=[UniProtKB:P55000], key=n01)",
	    "add_qedge(subject=n00, object=n01, key=e00)",
	    "expand(edge_key=e00,kp=RTX-KG2)",
	    "resultify()",
	    "return(message=true, store=false)"
	]
    },
    "submitter": "sramsey via curl script (working on issue 1754)",
    "stream_progress": true
}
```

**Comment by saramsey** [comment]
Here are the results of running Workflow A through the ARAX and RTX-KG2 services (using the code in the RTX "production" branch) in the "production" (and, via Expand, the local "kg2" dev-area) on the `arax-backup.rtx.ai` instance:
```
[
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct_inverse.json",
        "status_code": 200,
        "result_count": 4,
        "timedelta": "0:00:09.849588"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct.json",
        "status_code": 200,
        "result_count": 2,
        "timedelta": "0:00:09.164961"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.2_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:02:37.830905"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.2a_expanded_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 57,
        "timedelta": "0:00:17.739659"
    }
]
```

**Comment by saramsey** [comment]
Results running against `arax.ncats.io`:
```
[
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct_inverse.json",
        "status_code": 200,
        "result_count": 4,
        "timedelta": "0:00:09.401267"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct.json",
        "status_code": 200,
        "result_count": 2,
        "timedelta": "0:00:09.049903"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.2_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:02:12.360804"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.2a_expanded_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 57,
        "timedelta": "0:00:13.623891"
    }
]
```

**Comment by saramsey** [comment]
OK, I have determined that the local "production" ARAX on `arax-backup.rtx.ai` was using KG2.7.3, which kind of invalidates the test, in my view. I am working on installing KG2.7.4 on `arax-backup.rtx.ai`.

**Comment by saramsey** [comment]
Results from running the script that runs workflows A, B, and C against the ARAX on `arax-backup.rtx.ai`:
```
[
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct_inverse.json",
        "status_code": 200,
        "result_count": 4,
        "timedelta": "0:00:09.349294"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct.json",
        "status_code": 200,
        "result_count": 2,
        "timedelta": "0:00:09.095879"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.2_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:04:29.489499"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "A.2a_expanded_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 57,
        "timedelta": "0:00:17.546676"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.3a_DILI-two-hop-chemical-entity.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:17:29.640524"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2e_DILI-fourth-one-hop-from-PUBCHEM.COMPOUND_5865_Prednisone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:53.466859"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2b_DILI-fourth-one-hop-from-MESH_D000077185_Resveratrol.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:36.003846"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2c_DILI-fourth-one-hop-from-MESH_D000077385_Silybin.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:34.447075"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2a_DILI-fourth-one-hop-from-CHEBI_41879_Dexamethasone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:48.018624"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2f_DILI-fourth-one-hop-from-CHEMBL.COMPOUND_CHEMBL50_Quercetin.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:37.304941"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.1_DILI-three-hop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:11:28.621694"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.3b_DILI-two-hop-drug.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:17:25.182875"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "B.2d_DILI-fourth-one-hop-from-PUBCHEM.COMPOUND_5877_Methylprednisone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:49.461212"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "C3.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:05:03.216575"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "C1.json",
        "status_code": 200,
        "result_count": 360,
        "timedelta": "0:00:36.184212"
    },
    {
        "endpoint": "http://localhost:5000/api/arax/v1.2/query",
        "query": "C2.json",
        "status_code": 200,
        "result_count": 40,
        "timedelta": "0:05:39.176852"
    }
]
```

**Comment by saramsey** [comment]
Results from running workflows A, B, and C against the ARAX on `arax.ncats.io` in the `production` dev-area:
```
[
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct_inverse.json",
        "status_code": 200,
        "result_count": 4,
        "timedelta": "0:00:08.417852"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.0_RHOBTB2_direct.json",
        "status_code": 200,
        "result_count": 2,
        "timedelta": "0:00:08.876546"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.2_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:02:35.136842"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "A.2a_expanded_RHOBTB2_twohop.json",
        "status_code": 200,
        "result_count": 57,
        "timedelta": "0:00:17.265293"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.3a_DILI-two-hop-chemical-entity.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:18:15.817551"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2e_DILI-fourth-one-hop-from-PUBCHEM.COMPOUND_5865_Prednisone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:56.655926"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2b_DILI-fourth-one-hop-from-MESH_D000077185_Resveratrol.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:36.963599"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2c_DILI-fourth-one-hop-from-MESH_D000077385_Silybin.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:35.931920"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2a_DILI-fourth-one-hop-from-CHEBI_41879_Dexamethasone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:50.559213"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2f_DILI-fourth-one-hop-from-CHEMBL.COMPOUND_CHEMBL50_Quercetin.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:38.808611"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.1_DILI-three-hop.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:11:26.581144"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.3b_DILI-two-hop-drug.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:17:53.446880"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "B.2d_DILI-fourth-one-hop-from-PUBCHEM.COMPOUND_5877_Methylprednisone.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:00:52.495491"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "C3.json",
        "status_code": 200,
        "result_count": 500,
        "timedelta": "0:05:12.426959"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "C1.json",
        "status_code": 200,
        "result_count": 360,
        "timedelta": "0:00:44.698166"
    },
    {
        "endpoint": "https://arax.ncats.io/api/arax/v1.2/query",
        "query": "C2.json",
        "status_code": 200,
        "result_count": 40,
        "timedelta": "0:05:39.848785"
    }
]
```

**Comment by saramsey** [comment]
Diffing the two sets of results, dropping differences in endpoint names and timestamps:
```
]rt@5d5ebba937d1:/mnt/data/orangeboard/issue-1754/minihackathons/scripts$ diff output_results_backup_prod.json save/output_results_real_prod.json | grep -v timedelta | grep -v endpoint
3c3
---
7c7
---
10c10
---
14c14
---
17c17
---
21c21
---
24c24
---
28c28
---
31c31
---
35c35
---
38c38
---
42c42
---
45c45
---
49c49
---
52c52
---
56c56
---
59c59
---
63c63
---
66c66
---
70c70
---
73c73
---
77c77
---
80c80
---
84c84
---
87c87
---
91c91
---
94c94
---
98c98
---
101c101
---
105c105
---
108c108
---
112c112
---
```
so the results appear to be the same, at least in terms of numbers of results returned for each sub-workflow query.

**Comment by saramsey** [comment]
Now, I will show diffs of the output files for running workflows A, B, and C through the  `arax-backup.rtx.ai` ARAX endpoint using the `production` code and using the `production` code plus PR1748. No difference:
```
ubuntu@ip-172-31-48-139:/mnt/data/orangeboard/issue-1754/minihackathons/scripts$ diff output_results_backup_prod.json output_results_backup_prod_PR1748.json | grep -v time
7c7
---
14c14
---
21c21
---
28c28
---
35c35
---
42c42
---
49c49
---
56c56
---
63c63
---
70c70
---
77c77
---
84c84
---
91c91
---
98c98
---
105c105
---
112c112
---
```

**Comment by saramsey** [comment]
Checking for zombie processes:
```
ubuntu@ip-172-31-48-139:/mnt/data/orangeboard/issue-1754/minihackathons/scripts$ sudo docker exec rtx1 ps axwf
  PID TTY      STAT   TIME COMMAND
 2997 ?        Rs     0:00 ps axwf
 2655 pts/3    Ss     0:00 bash
 2770 pts/3    S      0:00  \_ su - rt
 2771 pts/3    S+     0:00      \_ -su
 2065 pts/4    Ss     0:00 bash
 2344 pts/4    S      0:00  \_ su - rt
 2345 pts/4    S+     0:00      \_ -su
 1475 pts/1    Ss+    0:00 bash
  996 pts/2    Ss     0:00 bash
 1011 pts/2    S      0:00  \_ su - rt
 1012 pts/2    S+     0:00      \_ -su
    1 pts/0    Ss+    0:00 bash
   38 ?        Ss     0:01 /usr/sbin/apache2 -k start
   41 ?        Sl     0:13  \_ /usr/sbin/apache2 -k start
   42 ?        Sl     0:13  \_ /usr/sbin/apache2 -k start
 2709 ?        S      1:15 python3 -u -m openapi_server
 2747 ?        S      0:35 python3 -u -m openapi_server
```

**Comment by saramsey** [comment]
I believe PR1748 is safe to merge into the "production" branch.

**Comment by saramsey** [comment]
OK, I think the commits for the zombie bug fix (#1744) are ready to merge to the "production" branch and to deploy to arax.ncats.io.  Any concerns or objections?

**Comment by edeutsch** [comment]
Do you have approval from all the workflow owners? I don't see that anywhere. I am fine to merge, but the protocol is that all workflow owners are supposed to sign off on any changes.

I notice you don't mention workflow D in previous comment?

**Comment by saramsey** [comment]
Hi Eric, thanks for your reply.  I didn't test Workflow D, in light of #1755.

To solicit approval from the workflow owners, do you know what is the approved/sensible procedure?  DM them?  Post on the general slack channel in the NCATSTranslator workspace?  Please pardon my ignorance. If DM, is there a list of the owners that I can consult to know whom to DM?

**Comment by edeutsch** [comment]
> Hi Eric, thanks for your reply. I didn't test Workflow D, in light of #1755. But now, looking at #1755 more closely, I think I _can_ test as long as I don't use `/asyncquery`, right?

correct. Plus there's also D.1, D.2, D.3 that aren't a problem?

> To solicit approval from the workflow owners, do you know what is the approved/sensible procedure? DM them? Post on the general slack channel in the NCATSTranslator workspace? If DM, is there a list of the owners that I can consult to know whom to DM? Please pardon my ignorance.

See the Translator #general channel. Scroll back to December 3 and start reading.

**Comment by saramsey** [comment]
OK, @edeutsch from your suggestion, I discern that 

- C. Bizon is an owner for Workflow A
- K. Fecho is an owner for Workflow B
- J. Hadlock is an owner for Workflow C
- B. Smith is an owner for Workflow D


anything I am missing?

**Comment by saramsey** [comment]
I am testing Workflow D now on `arax-backup.rtx.ai`

**Comment by edeutsch** [comment]
I infer the same based on my reading of the #general convo, yes.

If you're planning on tagging them, I would also tag Mark Williams since he's going to be the Demo Driver and probably has the most skin in the game..

**Comment by saramsey** [comment]
> I infer the same based on my reading of the #general convo, yes.
> 
> If you're planning on tagging them, I would also tag Mark Williams since he's going to be the Demo Driver and probably has the most skin in the game..

Thank you, that is helpful information. Will do.

**Comment by saramsey** [comment]
Confirmed-- the patch in [PR 1748](https://github.com/RTXteam/RTX/pull/1748) has no effect on the results for Workflow D:
```
rt@5d5ebba937d1:/mnt/data/orangeboard/issue-1754/minihackathons/scripts$ diff output_results_backup_prod-D.json  output_results_backup_prod_PR1748-D.json | grep -v timedelta
7c7
---
14c14
---
21c21
---
28c28
---
35c35
---
```

**Comment by saramsey** [comment]
Owners for Workflows A, C, and D have approved. Waiting on approval from the Workflow B owner.

**Comment by MarkDWilliams** [comment]
Provided there isn't an impact on the outcome of the workflows, which it looks like there isn't, I don't have any objections on my end.

**Comment by saramsey** [comment]
Workflow B owners have also approved.

**Comment by saramsey** [comment]
I am going to deploy this today.

**Comment by saramsey** [comment]
Since we are going to eventually update the `Dec2021Demo` tag to point to the new head of the `production` branch (right?), I have created a back-up tag `Dec2021Demo-20211116` in case we need to revert to the previous tagged state of the `production` branch. You can see that, at the moment, the two reference _identical_ commits in the `production` branch:



<img width="416" alt="Screen Shot 2021-12-07 at 11 23 44 AM" src="https://user-images.githubusercontent.com/5562107/145093142-6da9d205-bd78-4be2-b9f4-15959d9e47eb.png">

**Comment by saramsey** [comment]
To be absolutely sure that `Dec2021Demo` is the same as the current tip of the `production` branch, I made a branch `issue-1754` from a checkout of the `Dec2021Demo` tag, and compared it to the `production` branch:

<img width="1028" alt="Screen Shot 2021-12-07 at 11 34 20 AM" src="https://user-images.githubusercontent.com/5562107/145094423-d8b672c0-aa9c-40dd-b1ad-484325dcbaf7.png">

**Comment by saramsey** [comment]
I've pushed the code patch into the `production` branch using commits 365f7a4 and f0eea4d, as shown here with a diff of `production` against the baseline `issue-1754` branch:

<img width="956" alt="Screen Shot 2021-12-07 at 11 43 36 AM" src="https://user-images.githubusercontent.com/5562107/145095514-cee61872-3722-4afe-a481-d92183368835.png">

**Comment by saramsey** [comment]
You can see the start of the patch, in `RTX/code/UI/OpenAPI/python-flask-server/openapi_server/__main__.py` in the `production` branch (note the `while` loop):

<img width="1000" alt="Screen Shot 2021-12-07 at 11 44 42 AM" src="https://user-images.githubusercontent.com/5562107/145095762-3220890a-edf7-4295-b404-3868aa1a9cf4.png">

**Comment by saramsey** [comment]
Same goes for `RTX/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/__main__.py` in the `production` branch, note the `while` loop:

<img width="967" alt="Screen Shot 2021-12-07 at 11 45 52 AM" src="https://user-images.githubusercontent.com/5562107/145095870-8a76ebe3-928b-4bac-93c9-a4d45b023408.png">

**Comment by saramsey** [comment]
Having demonstrated that the code patch (and _only_ the code patch) was deployed into the `production` branch, we are ready to deploy it to `arax-backup.rtx.ai`.

**Comment by saramsey** [comment]
Test of the "hello, world" acetaminophen query on `arax-backup.rtx.ai` worked:

<img width="827" alt="Screen Shot 2021-12-07 at 11 54 25 AM" src="https://user-images.githubusercontent.com/5562107/145096994-c96a8651-d1e6-4840-93f3-8a15997b2219.png">

**Comment by saramsey** [comment]
I have updated the code in the `/mnt/data/orangeboard/production/RTX` repo inside the `rtx1` container on `arax.ncats.io`:
```
rt@d1fd345478a0:/mnt/data/orangeboard/production/RTX$ git pull origin production
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 13 (delta 9), reused 13 (delta 9), pack-reused 0
Unpacking objects: 100% (13/13), done.
From https://github.com/RTXTeam/RTX
 * branch            production -> FETCH_HEAD
   3cf4938..f0eea4d  production -> origin/production
Updating 3cf4938..f0eea4d
Fast-forward
 code/UI/OpenAPI/python-flask-server/KG2/openapi_server/__main__.py | 21 ++++++++++++++++-----
 code/UI/OpenAPI/python-flask-server/openapi_server/__main__.py     | 21 ++++++++++++++++-----
```
and in the `/mnt/data/orangeboard/kg2/RTX` repo as well:
```
rt@d1fd345478a0:/mnt/data/orangeboard/kg2/RTX$ git pull origin production
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 13 (delta 9), reused 13 (delta 9), pack-reused 0
Unpacking objects: 100% (13/13), done.
From https://github.com/RTXteam/RTX
 * branch            production -> FETCH_HEAD
   3cf4938..f0eea4d  production -> origin/production
Updating 3cf4938..f0eea4d
Fast-forward
 code/UI/OpenAPI/python-flask-server/KG2/openapi_server/__main__.py | 21 ++++++++++++++++-----
 code/UI/OpenAPI/python-flask-server/openapi_server/__main__.py     | 21 ++++++++++++++++-----
 2 files changed, 32 insertions(+), 10 deletions(-)
```

**Comment by saramsey** [comment]
I am planning to restart `RTX_OpenAPI_kg2` and `RTX_OpenAPI_production` on `arax.ncats.io` inside the `rtx1` container, at 21:30 PST Tue Dec. 7.

**Comment by saramsey** [comment]
OK, deployed to `arax.ncats.io`.  Closing out this issue.
